500 error because of database lock

## Specify type:
- Bug

### Bug severity (if applicable):
- Medium
___

## Description:
If requests are timed just right, it is possible to get a 500 error when attempting to pause or cancel a scan. NOTE: This has been observed with the pilot container, but probably is still an issue in master.

## Bug Report

### Expected behavior:
If I make a poorly timed request because some other task is accessing the object I'm attempting to modify, the server does something graceful.

### Actual behavior:
The server tells me it is broken.

For the following satellite source:
```
 'source_type': 'satellite'}]
source = {'credentials': ['sat6'], 'hosts': ['sat-r220-07.lab.eng.rdu2.redhat.com'], 'name': 'sat6', 'options': {'satellite_version': '6.2', 'ssl_cert_verify': False}}
scan_type = 'inspect'
```

After the scan has started, if I attempt to pause it, I get the following 500:
```
E           request path : '/api/v1/scans/29/pause/'
E           
E           request body : b'{}'
E           
E           request headers : {'User-Agent': 'python-requests/2.18.4', 'Accept-Encoding': 'gzip, deflate', 'Accept': '*/*', 'Connection': 'keep-alive', 'Authorization': 'Token 2c1aa5e4d44ec1069c25982a20375550b808553d', 'Content-Length': '2', 'Content-Type': 'application/json'}
E           
E           response code : 500
E           
E           text_error_message : ('OperationalError at /api/v1/scans/29/pause/\n'
E            'database is locked\n'
E            '\n'
E            'Request Method: PUT\n'
E            'Request URL: https://127.0.0.1/api/v1/scans/29/pause/\n'
E            'Django Version: 1.11.5\n'
E            'Python Executable: /root/venv/bin/python3\n'
E            'Python Version: 3.6.4\n'
```

### Steps to reproduce:
I've not been able to reproduce manually yet, but this is happening with some regularity on one of our automation jobs -- I can point you to the logs.

### Environment information:
Provide detailed information on the *scanning* and *target* systems (if applicable); see example below.

Type            | Operating System | OS Version | Python Version | Virtualization
--------------- | ---------------- | ---------- | -------------- | --------------
Scanning System | RHEL             | 7.4        | 3.6            | Bare Metal
Target System   | RHEL             | 5.9        | 2.4.6          | RHEV
quipucords | docker container | pilot | |

--- a/quipucords/api/scanjob/model.py
+++ b/quipucords/api/scanjob/model.py
@@ -15,7 +15,7 @@ These models are used in the REST definitions.
 from datetime import datetime
 import logging
 from django.utils.translation import ugettext as _
-from django.db import (models, transaction)
+from django.db import (models, transaction, OperationalError)
 from django.db.models import Q
 from api.source.model import Source
 from api.scan.model import Scan, ScanOptions
@@ -194,8 +194,32 @@ class ScanJob(models.Model):
                             self.start_time).total_seconds()
         return elapsed_time
 
-    @transaction.atomic
+    def safe_state_change(self, method_name, message=None):
+        """Update state in protected clause."""
+        # django.db.utils.OperationalError: database is locked
+        try:
+            method = getattr(self, method_name)
+            if method_name == '_fail':
+                method(message)
+            else:
+                method()
+        except OperationalError:
+            # try one more time
+            method = getattr(self, method_name)
+            if method_name == '_fail':
+                method(message)
+            else:
+                method()
+
     def queue(self):
+        """Queue the job to run.
+
+        Change job state from CREATED TO PENDING.
+        """
+        self.safe_state_change('_queue')
+
+    @transaction.atomic
+    def _queue(self):
         """Queue the job to run.
 
         Change job state from CREATED TO PENDING.
@@ -291,8 +315,15 @@ class ScanJob(models.Model):
         self.save()
         self.log_current_status()
 
-    @transaction.atomic
     def start(self):
+        """Start a job.
+
+        Change job state from PENDING TO RUNNING.
+        """
+        self.safe_state_change('_start')
+
+    @transaction.atomic
+    def _start(self):
         """Start a job.
 
         Change job state from PENDING TO RUNNING.
@@ -309,8 +340,18 @@ class ScanJob(models.Model):
         self.save()
         self.log_current_status()
 
-    @transaction.atomic
     def restart(self):
+        """Restart a job.
+
+        Change job state from PENDING, PAUSED OR RUNNING
+        TO RUNNING status. Handles job that need to be run,
+        jobs that were running and the server quick, and
+        paused jobs.
+        """
+        self.safe_state_change('_restart')
+
+    @transaction.atomic
+    def _restart(self):
         """Restart a job.
 
         Change job state from PENDING, PAUSED OR RUNNING
@@ -336,8 +377,15 @@ class ScanJob(models.Model):
         self.save()
         self.log_current_status()
 
-    @transaction.atomic
     def pause(self):
+        """Pause a job.
+
+        Change job state from PENDING/RUNNING TO PAUSED.
+        """
+        self.safe_state_change('_pause')
+
+    @transaction.atomic
+    def _pause(self):
         """Pause a job.
 
         Change job state from PENDING/RUNNING TO PAUSED.
@@ -362,8 +410,16 @@ class ScanJob(models.Model):
         self.save()
         self.log_current_status()
 
-    @transaction.atomic
     def cancel(self):
+        """Cancel a job.
+
+        Change job state from CREATED, PENDING, RUNNING, or
+        PAUSED to CANCELED.
+        """
+        self.safe_state_change('_cancel')
+
+    @transaction.atomic
+    def _cancel(self):
         """Cancel a job.
 
         Change job state from CREATED, PENDING, RUNNING, or
@@ -393,8 +449,15 @@ class ScanJob(models.Model):
         self.save()
         self.log_current_status()
 
-    @transaction.atomic
     def complete(self):
+        """Complete a job.
+
+        Change job state from RUNNING TO COMPLETE.
+        """
+        self.safe_state_change('_complete')
+
+    @transaction.atomic
+    def _complete(self):
         """Complete a job.
 
         Change job state from RUNNING TO COMPLETE.
@@ -412,8 +475,16 @@ class ScanJob(models.Model):
         self._log_stats('COMPLETION STATS.')
         self.log_current_status()
 
-    @transaction.atomic
     def fail(self, message):
+        """Fail a job.
+
+        Change job state from RUNNING TO COMPLETE.
+        :param message: The error message associated with failure
+        """
+        self.safe_state_change('_fail', message)
+
+    @transaction.atomic
+    def _fail(self, message):
         """Fail a job.
 
         Change job state from RUNNING TO COMPLETE.

--- a/quipucords/api/scanjob/view.py
+++ b/quipucords/api/scanjob/view.py
@@ -218,8 +218,9 @@ class ScanJobViewSet(mixins.RetrieveModelMixin,
             raise ValidationError(error)
         scan = get_object_or_404(self.queryset, pk=pk)
         if scan.status == ScanTask.RUNNING:
-            scan.pause()
+            # Kill job before changing job state
             pause_scan.send(sender=self.__class__, instance=scan)
+            scan.pause()
             serializer = ScanJobSerializer(scan)
             json_scan = serializer.data
             json_scan = expand_scanjob(json_scan)
@@ -246,8 +247,9 @@ class ScanJobViewSet(mixins.RetrieveModelMixin,
             err_msg = _(messages.NO_CANCEL)
             return JsonResponse({'non_field_errors': [err_msg]}, status=400)
 
-        scan.cancel()
+        # Kill job before changing job state
         cancel_scan.send(sender=self.__class__, instance=scan)
+        scan.cancel()
         serializer = ScanJobSerializer(scan)
         json_scan = serializer.data
         json_scan = expand_scanjob(json_scan)
@@ -263,6 +265,7 @@ class ScanJobViewSet(mixins.RetrieveModelMixin,
             raise ValidationError(error)
         scan = get_object_or_404(self.queryset, pk=pk)
         if scan.status == ScanTask.PAUSED:
+            # Update job state before starting job
             scan.restart()
             restart_scan.send(sender=self.__class__,
                               instance=scan)

